singlem/condense.py

```diff
@@ -0,0 +1,297 @@
+                        excluded_markers.add(gene)
+                    continue
+                tax_split = taxonomy.split('; ')
+                # ensure OTU is assigned to the domain level or higher
+                if len(tax_split) < 2:  # contains at least Root; d__DOMAIN
+                    continue
+                domain = tax_split[1].strip('d__')
+                # ensure the domain of this OTU is targeted by the gene
+                if domain not in markers[gene]: 
+                    continue
+                if sample not in samples:
+                    samples[sample] = {}
+                    taxon_counter[sample] = {}
+                    queues[sample] = PriorityQueue()
+                if gene not in marker_set:
+                    marker_set.add(gene)
+                if gene not in samples[sample]:
+                    # create new tree for this marker
+                    samples[sample][gene] = WordNode(None, "Root")
+                # this taxonomy has been added already?
+                taxbuilder = "Root"
+                priority = 0
+                # process taxonomy levels in ascending order
+                for tax in tax_split[1:]:
+                    taxbuilder += "; " + tax
+                    priority += 1
+                    if taxbuilder not in taxon_counter[sample]:
+                        queues[sample].put((priority, taxbuilder))
+                        taxon_counter[sample][taxbuilder] = set()
+                # add to tree
+                samples[sample][gene].add_words(tax_split, coverage)
+                taxon_counter[sample][taxonomy].add(gene)
+        
+        for gene in markers:
+            if gene not in marker_set:
+                logging.warning("Marker %s not found in OTU table, removing..." % gene)
+                for domain in markers[gene]:
+                    target_domains[domain] -= 1
+                    if target_domains[domain] in [1, 2]:
+                        raise Exception("Number of markers for all domains must either be >= 3 or equal to 0. Only {} markers for domain '{}' found".format(target_domains[domain], domain))
+        
+        ### Trim the taxonomy coverages
+        logging.info("Trimming constructed taxonomy trees.")
+        for sample in queues:
+            while not queues[sample].empty():
+                taxonomy = queues[sample].get()[1] # tuple: (priority, taxonomy)
+                tax_split = taxonomy.split("; ")
+                domain = tax_split[1].strip("d__")
+                genes = {}
+                for gene in samples[sample]:
+                    node = samples[sample][gene].get_tree(taxonomy.split("; "))
+                    # node returns False if it does not exist
+                    if node:
+                        cov = node.get_full_coverage()
+                        genes[gene] = cov
+                # do not trim if taxonomy only remains in 2 markers or less
```

Comment:
The assumption this is making is that the lineage is probably correct up to a certain level, but is incorrect at the higher resolutions. So instead of just removing the entry entirely, the coverage simply gets shifted down to a taxonomy level that contains coverage at least 3 markers or barring that, just places it at the domain level.